### DISCUSSION TITLE
Renovate not updating relocated dependencies

### DISCUSSION THREAD

**Discussion by NotMyFault** [renovatebot/renovate#8104] [Request Help]
Heyo,

I was wondering if renovate requires an additional config entry to update relocated/shaded dependencies.
Renovate raised [this](https://github.com/IntellectualSites/FastAsyncWorldEdit/pull/800/files) pull request and updated the version from 1.7 to 1.8. However, the version of the shadow jar task has either been ignored or not recognized, as it is still 1.7 in the pr named.
Is there anything I forgot?

Cheers

**Comment by rarkins** [comment]
> the version of the shadow jar task

Can you clarify what/where exactly this is?

**Reply by NotMyFault** [reply]
> > the version of the shadow jar task
> 
> Can you clarify what/where exactly this is?

For sure. Renovate bumped the version for the compile scope [here](https://github.com/IntellectualSites/FastAsyncWorldEdit/pull/800/files#diff-f70bd30806cc6e50ad99b03e389e53ffc563052e0de0707b0c3faae0a6a662f9R71) but the version [being shaded](https://github.com/IntellectualSites/FastAsyncWorldEdit/pull/800/files#diff-f70bd30806cc6e50ad99b03e389e53ffc563052e0de0707b0c3faae0a6a662f9R112) hasn't been touched.
Line 112 of the same file, incase the link doesn't do its thing.

**Reply by viceice** [reply]
I think [this](https://github.com/IntellectualSites/FastAsyncWorldEdit/blob/d77e16c66fceb3600a492fddffac5b47178dc364/worldedit-bukkit/build.gradle.kts#L112)  is currently not possible, you can try our new [gradle-lite](https://docs.renovatebot.com/modules/manager/gradle-lite/) manager, maybe it can find that dep.

**Reply by NotMyFault** [reply]
> I think [this](https://github.com/IntellectualSites/FastAsyncWorldEdit/blob/d77e16c66fceb3600a492fddffac5b47178dc364/worldedit-bukkit/build.gradle.kts#L112) is currently not possible, you can try our new [gradle-lite](https://docs.renovatebot.com/modules/manager/gradle-lite/) manager, maybe it can find that dep.

Unfortunately it still targets the `compile` one.

**Reply by viceice** [reply]
Can you please provide a minimal reproduction repo,so we can debug this issue.

**Reply by NotMyFault** [reply]
> Can you please provide a minimal reproduction repo,so we can debug this issue.

I've created https://github.com/N0tMyFaultOG/renovate-dependency-test. Just waiting for renovate to raise the pull request.

**Reply by NotMyFault** [reply]
> > Can you please provide a minimal reproduction repo,so we can debug this issue.
> 
> I've created https://github.com/N0tMyFaultOG/renovate-dependency-test. Just waiting for renovate to raise the pull request.

Alright, renovate raised the pr. [Looking at the files changed](https://github.com/N0tMyFaultOG/renovate-dependency-test/pull/1/files) only the dependency in the `dependencies { }` block has been targeted. bStats is being shaded [at line 45](https://github.com/N0tMyFaultOG/renovate-dependency-test/pull/1/files#diff-c0dfa6bc7a8685217f70a860145fbdf416d449eaff052fa28352c5cec1a98c06R45) but that value has been left untouched.
Here's my [renovate.json](https://github.com/N0tMyFaultOG/renovate-dependency-test/blob/main/renovate.json) used for the run.

**Reply by viceice** [reply]
can you paste the json after `packageFiles with updates` debug message

**Reply by NotMyFault** [reply]
> can you paste the json after `packageFiles with updates` debug message

Here you go https://gist.github.com/N0tMyFaultOG/3995864452533f255a658162f19a3b9f, took me a minute to find it.

**Reply by viceice** [reply]
![image](https://user-images.githubusercontent.com/1798109/102881900-d6180100-444d-11eb-99c9-80c16d0cc8ce.png)

gradle-lite missed to add registry urls, so lookup failed. Can you assign registryUrl manually to `org.bstats:bstats-bukkit` with package rule?

**Reply by NotMyFault** [reply]
For sure. I'm pretty new to renovate, mind giving me an example?

**Reply by viceice** [reply]
Try add this to `renovate.json`
```json
"packageRules": [
    {
      "packageNames": ["org.bstats:bstats-bukkit"],
      "registryUrls": [
              "https://papermc.io/repo/repository/maven-public/",
              "https://repo.codemc.org/repository/maven-public",
              "https://plugins.gradle.org/m2"
            ]
    }
]
```

**Reply by NotMyFault** [reply]
Now it's been detected properly in both spots: https://github.com/N0tMyFaultOG/renovate-dependency-test/pull/2/files

**Reply by viceice** [reply]
you can now add 

```json
  "gradle": {
    "enabled": false
  }
```

to disable default gradle manager, that will speedup renovate runs

**Reply by viceice** [reply]
maybe you need to play with package rule to assign registry urls to all gradle packages, eg:

```json
"packageRules": [
    {
      "managers": ["gradle-lite"],
      "registryUrls": [
              "https://papermc.io/repo/repository/maven-public/",
              "https://repo.codemc.org/repository/maven-public",
              "https://plugins.gradle.org/m2"
            ]
    }
]
```

**Reply by NotMyFault** [reply]
I've added the following and it detects both occurrences still https://github.com/N0tMyFaultOG/renovate-dependency-test/pull/3/files

**Reply by viceice** [reply]
@zharinov Can you look at the `registryUrls` issue ? Looks like we can extract them too with `gradle-lite` from here:
https://github.com/N0tMyFaultOG/renovate-dependency-test/blob/b89b9376642ab68c25c13d50d85cab7f0d37eaa9/build.gradle.kts#L26

**Reply by zharinov** [reply]
Just extended the parser [here](https://github.com/renovatebot/renovate/pull/8120), though not sure about other failed lookups (e.g. `com.destroystokyo.paper:paper-api`)

<details><summary>Extracted config with updates</summary>
<p>

```json
{
  "gradle-lite": [
    {
      "packageFile": "build.gradle.kts",
      "datasource": "maven",
      "deps": [
        {
          "depType": "plugin",
          "depName": "org.cadixdev.licenser",
          "lookupName": "org.cadixdev.licenser:org.cadixdev.licenser.gradle.plugin",
          "registryUrls": [
            "https://plugins.gradle.org/m2/",
            "https://papermc.io/repo/repository/maven-public/",
            "https://repo.codemc.org/repository/maven-public"
          ],
          "currentValue": "0.5.0",
          "commitMessageTopic": "plugin org.cadixdev.licenser",
          "managerData": {
            "fileReplacePosition": 303,
            "packageFile": "build.gradle.kts"
          },
          "depIndex": 0,
          "updates": [],
          "warnings": [
            {
              "depName": "org.cadixdev.licenser",
              "message": "Failed to look up dependency org.cadixdev.licenser"
            }
          ],
          "fixedVersion": "0.5.0"
        },
        {
          "depType": "plugin",
          "depName": "net.minecrell.plugin-yml.bukkit",
          "lookupName": "net.minecrell.plugin-yml.bukkit:net.minecrell.plugin-yml.bukkit.gradle.plugin",
          "registryUrls": [
            "https://plugins.gradle.org/m2/",
            "https://papermc.io/repo/repository/maven-public/",
            "https://repo.codemc.org/repository/maven-public"
          ],
          "currentValue": "0.3.0",
          "commitMessageTopic": "plugin net.minecrell.plugin-yml.bukkit",
          "managerData": {
            "fileReplacePosition": 361,
            "packageFile": "build.gradle.kts"
          },
          "depIndex": 1,
          "updates": [],
          "warnings": [
            {
              "depName": "net.minecrell.plugin-yml.bukkit",
              "message": "Failed to look up dependency net.minecrell.plugin-yml.bukkit"
            }
          ],
          "fixedVersion": "0.3.0"
        },
        {
          "depType": "plugin",
          "depName": "com.github.johnrengelman.shadow",
          "lookupName": "com.github.johnrengelman.shadow:com.github.johnrengelman.shadow.gradle.plugin",
          "registryUrls": [
            "https://plugins.gradle.org/m2/",
            "https://papermc.io/repo/repository/maven-public/",
            "https://repo.codemc.org/repository/maven-public"
          ],
          "currentValue": "6.1.0",
          "commitMessageTopic": "plugin com.github.johnrengelman.shadow",
          "managerData": {
            "fileReplacePosition": 419,
            "packageFile": "build.gradle.kts"
          },
          "depIndex": 2,
          "updates": [],
          "warnings": [],
          "fixedVersion": "6.1.0",
          "dependencyUrl": "com/github/johnrengelman/shadow/com.github.johnrengelman.shadow.gradle.plugin"
        },
        {
          "depType": "plugin",
          "depName": "org.jetbrains.kotlin.jvm",
          "lookupName": "org.jetbrains.kotlin.jvm:org.jetbrains.kotlin.jvm.gradle.plugin",
          "registryUrls": [
            "https://plugins.gradle.org/m2/",
            "https://papermc.io/repo/repository/maven-public/",
            "https://repo.codemc.org/repository/maven-public"
          ],
          "currentValue": "1.4.21",
          "commitMessageTopic": "plugin org.jetbrains.kotlin.jvm",
          "managerData": {
            "fileReplacePosition": 453,
            "packageFile": "build.gradle.kts"
          },
          "depIndex": 3,
          "updates": [],
          "warnings": [],
          "fixedVersion": "1.4.21",
          "dependencyUrl": "org/jetbrains/kotlin/jvm/org.jetbrains.kotlin.jvm.gradle.plugin"
        },
        {
          "depName": "com.destroystokyo.paper:paper-api",
          "currentValue": "1.16.4-R0.1-SNAPSHOT",
          "managerData": {
            "fileReplacePosition": 966,
            "packageFile": "build.gradle.kts"
          },
          "registryUrls": [
            "https://papermc.io/repo/repository/maven-public/",
            "https://repo.codemc.org/repository/maven-public"
          ],
          "depIndex": 4,
          "updates": [],
          "warnings": [
            {
              "depName": "com.destroystokyo.paper:paper-api",
              "message": "Failed to look up dependency com.destroystokyo.paper:paper-api"
            }
          ],
          "fixedVersion": "1.16.4-R0.1-SNAPSHOT"
        },
        {
          "depName": "org.bstats:bstats-bukkit",
          "currentValue": "1.7",
          "managerData": {
            "fileReplacePosition": 1034,
            "packageFile": "build.gradle.kts"
          },
          "registryUrls": [
            "https://papermc.io/repo/repository/maven-public/",
            "https://repo.codemc.org/repository/maven-public"
          ],
          "depIndex": 5,
          "updates": [
            {
              "fromVersion": "1.7",
              "toVersion": "1.8",
              "newValue": "1.8",
              "newMajor": 1,
              "newMinor": 8,
              "updateType": "minor",
              "isSingleVersion": true
            }
          ],
          "warnings": [],
          "fixedVersion": "1.7",
          "dependencyUrl": "org/bstats/bstats-bukkit"
        },
        {
          "depName": "org.bstats:bstats-bukkit",
          "currentValue": "1.7",
          "managerData": {
            "fileReplacePosition": 1257,
            "packageFile": "build.gradle.kts"
          },
          "registryUrls": [
            "https://papermc.io/repo/repository/maven-public/",
            "https://repo.codemc.org/repository/maven-public"
          ],
          "depIndex": 6,
          "updates": [
            {
              "fromVersion": "1.7",
              "toVersion": "1.8",
              "newValue": "1.8",
              "newMajor": 1,
              "newMinor": 8,
              "updateType": "minor",
              "isSingleVersion": true
            }
          ],
          "warnings": [],
          "fixedVersion": "1.7",
          "dependencyUrl": "org/bstats/bstats-bukkit"
        }
      ]
    },
    {
      "packageFile": "settings.gradle.kts",
      "datasource": "maven",
      "deps": []
    }
  ]
}
```

</p>
</details>

**Reply by viceice** [reply]
`com.destroystokyo.paper:paper-api` couln't be find by normal `gradle` manager too

![image](https://user-images.githubusercontent.com/1798109/102907941-0758f700-4477-11eb-8144-c2fdd8954c97.png)


@N0tMyFaultOG Did you repo is missing any registry for that package ? Or is is a private package?

**Reply by NotMyFault** [reply]
Uh, odd. It's not my repository, but papermc and bStats are both public accessible. [paper-api](https://papermc.io/repo/#browse/browse:maven-snapshots:com%2Fdestroystokyo%2Fpaper%2Fpaper-api%2F1.16.4-R0.1-SNAPSHOT%2F1.16.4-R0.1-20201221.161441-77), [bstats-bukkit](https://repo.codemc.io/#browse/browse:maven-public:org%2Fbstats%2Fbstats-bukkit%2F1.8).

**Reply by zharinov** [reply]
Okay, then it seems like datasource-level issue

**Reply by zharinov** [reply]
To be more precise, it's about [404 errors](https://papermc.io/repo/repository/maven-public/com/destroystokyo/paper/paper-api/1.11-R0.1-SNAPSHOT/paper-api-1.11-R0.1-SNAPSHOT.pom) for POM lookups. Thus, `RENOVATE_EXPERIMENTAL_NO_MAVEN_POM_CHECK` should be set.

**Reply by NotMyFault** [reply]
Interesting, the 404 outlined refers to an artifact that is 4 years old, wondering that such an oldie is still relevant. Anyway, should I add the arg to my renovate.json or do you want to test it on your fork?

**Reply by zharinov** [reply]
Recently, we've introduced mandatory POM file presence check (by making HEAD request), in order to filter versions which are present in [`metadata.xml`](https://papermc.io/repo/repository/maven-snapshots/com/destroystokyo/paper/paper-api/maven-metadata.xml), but actual jars are missing in corresponding folder. It helps us to avoid broken upgrades/builds.

It works great for Maven central, but for some other repos there are false hits. In your case, that's because we look available versions here:
[.../com/destroystokyo/paper/paper-api/maven-metadata.xml](https://papermc.io/repo/repository/maven-snapshots/com/destroystokyo/paper/paper-api/maven-metadata.xml)
Then we try to check corresponding POMs like this:
[.../com/destroystokyo/paper/paper-api/1.16.4-R0.1-SNAPSHOT/paper-api-1.16.4-R0.1-SNAPSHOT.pom](https://papermc.io/repo/repository/maven-snapshots/com/destroystokyo/paper/paper-api/1.16.4-R0.1-SNAPSHOT/paper-api-1.16.4-R0.1-SNAPSHOT.pom)
But the actual snapshot POM located here:
[.../com/destroystokyo/paper/paper-api/1.16.4-R0.1-SNAPSHOT/paper-api-1.16.4-R0.1-20201221.161441-77.pom](https://papermc.io/repo/repository/maven-snapshots/com/destroystokyo/paper/paper-api/1.16.4-R0.1-SNAPSHOT/paper-api-1.16.4-R0.1-20201221.161441-77.pom)

BTW we already have [the issue](https://github.com/renovatebot/renovate/issues/3728) for that. But for quick workaround, `RENOVATE_EXPERIMENTAL_NO_MAVEN_POM_CHECK` environment variable also can be enabled.

**Reply by NotMyFault** [reply]
Oic, interesting. But that is only relevant to nested poms, isn't it? For the previous issue, would I need to use gradle-lite and add all repositories as `registryUrls` by hand?

**Reply by zharinov** [reply]
We already merged [the fix](https://github.com/renovatebot/renovate/pull/8120), so now (soon) gradle-lite should detect `papermc.io` and `codemc.org` repositories automatically

**Reply by rarkins** [reply]
@zharinov maybe we are better off *not* to do the HEAD by default and either:
- add ones we find to workarounds, and/or
- support the strict functionality by config (opt-in)
